backend/api/src/create-post.ts

```diff
@@ -78,89 +73,61 @@ export const createpost = authEndpoint(async (req, auth) => {
     }
   }
 
-  // if this is an advertisement, subtract amount from user's balance and write txn
-  let funds
-  if ('type' in otherProps && otherProps.type === 'ad') {
-    const cost = otherProps.totalCost
-
-    if (!cost) {
-      throw new APIError(403, 'totalCost required')
-    }
-
-    // deduct from user
-    await firestore.runTransaction(async (trans) => {
-      const result = await runTxn(trans, {
-        category: 'AD_CREATE',
-        fromType: 'USER',
-        fromId: creator.id,
-        toType: 'AD',
-        toId: postRef.id,
-        amount: cost,
-        token: 'M$',
-        description: 'Creating ad',
-      } as PostAdCreateTxn)
-      if (result.status == 'error') {
-        throw new APIError(500, result.message ?? 'An unknown error occurred')
-      }
-    })
-
-    // init current funds
-    funds = cost
-  }
-
   const post: Post = removeUndefinedProps({
     ...otherProps,
-    id: postRef.id,
+    id: randomUUID(),
     creatorId: creator.id,
     slug,
     title,
     isGroupAboutPost,
     createdTime: Date.now(),
     content: content,
     contractSlug,
-    funds,
     creatorName: creator.name,
     creatorUsername: creator.username,
     creatorAvatarUrl: creator.avatarUrl,
     visibility: 'public',
     groupId,
   })
 
-  await postRef.create(post)
+  // TODO: lock. or migrate groups.
+
   if (groupId) {
     const groupRef = firestore.collection('groups').doc(groupId)
     const group = await groupRef.get()
     if (group.exists) {
       const groupData = group.data()
       if (groupData) {
         const postIds = groupData.postIds ?? []
-        postIds.push(postRef.id)
+        postIds.push(post.id)
         await groupRef.update({ postIds })
-        await postRef.update({
-          visibility:
-            groupData.privacyStatus == 'private' ? 'private' : 'public',
-        })
+        post.visibility =
+          groupData.privacyStatus == 'private' ? 'private' : 'public'
       }
     }
   }
 
+  // currently uses the trigger to populate group_id, creator_id, created_time.
+  pg.none(`insert into posts (id, data) values ($1, $2)`, [post.id, post])
+
   return { status: 'success', post }
 })
 
 export const getSlug = async (title: string) => {
   const proposedSlug = slugify(title)
 
-  const preexistingPost = await getPostFromSlug(proposedSlug)
+  const preexistingPost = await postExists(proposedSlug)
 
   return preexistingPost ? proposedSlug + '-' + randomString() : proposedSlug
 }
 
-export async function getPostFromSlug(slug: string) {
-  const firestore = admin.firestore()
-  const snap = await firestore
-    .collection('posts')
-    .where('slug', '==', slug)
-    .get()
+// TODO: migrate slug in new column with unique constraint
+export async function postExists(slug: string) {
+  const pg = createSupabaseDirectClient()
```

Comment:
Creating the pg connection here is an anti-pattern. It should usually be initialized at the top level of the api endpoint or script or whatever, and then passed in as a parameter.